gobblin-core/src/main/java/gobblin/writer/http/SalesforceRestWriter.java

```diff
@@ -0,0 +1,372 @@
+   * Make it fail (throw exception) if status code is greater or equal to 400 except,
+   * the status code is 400 and error code is duplicate value, regard it as success(do not throw exception).
+   *
+   * If status code is 401 or 403, re-acquire access token before make it fail -- retry will take care of rest.
+   *
+   * {@inheritDoc}
+   * @see gobblin.writer.http.HttpWriter#processResponse(org.apache.http.HttpResponse)
+   */
+  @Override
+  public void processResponse(CloseableHttpResponse response) throws IOException, UnexpectedResponseException {
+    if (getLog().isDebugEnabled()) {
+      getLog().debug("Received response " + ToStringBuilder.reflectionToString(response, ToStringStyle.SHORT_PREFIX_STYLE));
+    }
+
+    int statusCode = response.getStatusLine().getStatusCode();
+    if (statusCode == 401 || statusCode == 403) {
+      getLog().info("Reacquiring access token.");
+      accessToken = null;
+      onConnect(oauthEndPoint);
+      throw new RuntimeException("Access denied. Access token has been reacquired and retry may solve the problem. "
+                                 + ToStringBuilder.reflectionToString(response, ToStringStyle.SHORT_PREFIX_STYLE));
+
+    }
+    if (batchSize > 1) {
+      processBatchRequestResponse(response);
+      numRecordsWritten += batchRecords.get().size();
+      batchRecords = Optional.absent();
+    } else {
+      processSingleRequestResponse(response);
+      numRecordsWritten++;
+    }
+  }
+
+  private void processSingleRequestResponse(CloseableHttpResponse response) throws IOException, UnexpectedResponseException {
+    int statusCode = response.getStatusLine().getStatusCode();
+    if (statusCode < 400) {
+      return;
+    }
+    String entityStr = EntityUtils.toString(response.getEntity());
+    if (statusCode == 400
+        && Operation.INSERT_ONLY_NOT_EXIST.equals(operation)
+        && response.getEntity() != null
+        && response.getEntity().getContent() != null) { //Ignore if it's duplicate entry error code
+      try {
+        JsonArray jsonArray = new JsonParser().parse(entityStr).getAsJsonArray();
+
+        JsonObject jsonObject = jsonArray.get(0).getAsJsonObject();
+        JsonElement val = jsonObject.get("errorCode");
+        if (val != null && val.getAsString().equals(DUPLICATE_VALUE_ERR_CODE)) {
+          return;
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("Unexpected response format. Response entity: " + entityStr, e);
+      }
+    }
+
+    throw new RuntimeException("Failed due to " + entityStr + " (Detail: " + ToStringBuilder.reflectionToString(response, ToStringStyle.SHORT_PREFIX_STYLE) + " )");
+  }
+
+  /**
+   * Check results from batch response, if any of the results is failure throw exception.
+   * @param response
+   * @throws IOException
+   * @throws UnexpectedResponseException
+   */
+  private void processBatchRequestResponse(CloseableHttpResponse response) throws IOException, UnexpectedResponseException {
```

Comment:
Removed duplicate on the logic to check if it's duplicate error code. 

(Not that much of common code mostly because all the information on batch response is serialized where only the error code part is serialized in single request response).